agent/consul/acl.go

```diff
@@ -1087,6 +1094,37 @@ func (f *aclFilter) allowSession(node string) bool {
 	return f.authorizer.SessionRead(node)
 }
 
+// allowEvent is used to determine if a stream.Event is accessible for an ACL.
+func (f *aclFilter) allowEvent(event stream.Event) bool {
+	// Fast path if ACLs are not enabled
+	if f.authorizer == nil {
+		return true
+	}
+
+	switch event.Topic {
+	case stream.Topic_ServiceHealth:
+		health := event.GetServiceHealth()
+		if health == nil {
+			return true
+		}
+		var node, service string
+		if health.ServiceNode != nil {
+			if health.ServiceNode.Node != nil {
+				node = health.ServiceNode.Node.Node
+			}
+			if health.ServiceNode.Service != nil {
+				service = health.ServiceNode.Service.Service
+			}
+		}
+		val := f.allowCheckServiceNode(node, service)
+		f.logger.Printf("checking allow on %s/%s (%v)", node, service, val)
+		return val
+	default:
+		f.logger.Printf("could not filter for unrecognized event topic %s", event.Topic)
+		return true
+	}
+}
+
```

Comment:
Hmm preferably type switching on specific event types should be as localized as possible. Otherwise implementing a new type requires code in lots of different places.

Could we maybe make this a method on `Event` so that this code is generic and all the type-specifics are contained in the Event file alongside the other things you'd have to change to add a new event type?

Reply:
Event is all generated so you'd have to go to two places anyway  (adding the proto declarations + acl logic file) - I put this here to reuse the `aclFilter` logic for existing things. I can move the aclFilter stuff to its own package and make it public to call it from the stream package if you think that's cleaner though.

Reply:
Ah yeah forgot it's generated.

Do you remember my idea about how we should have each event type be able to encode it's own set of ACL requirements so that we can un-ambiguously enforce them on client side later without compatibility issues etc? If we think along those lines then I think it suggest the same cleaner separation as I was getting at here:

 1. If possible, use some annotation in the protobuf and an extension or something to encode the ACL requirements by type (with `topic` and `key` as optional inputs?). If not possible, define a go file in the same place as the proto file that has a method that encodes this (so all are encoded in one place and any other type specialisation can go there too)
 2. have this method just call that method, retreive the set of things to enforce and validate that.

What do you think?

Reply:
Putting the ACL requirements in an explicit field makes sense so it's unambiguous to clients later - I just pushed up a change to do that.